File: dynabuffers-java/src/main/kotlin/dynabuffers/ast/datatype/OptionType.kt
package dynabuffers.ast.datatype

import dynabuffers.api.IRegistry
import dynabuffers.api.ISerializable
import dynabuffers.api.IType
import java.nio.ByteBuffer
import java.util.*

class OptionType(private val options: OptionTypeOptions) : IType, ISerializable {

    override fun size(value: Any?, registry: IRegistry): Int {
        if (!Objects.isNull(value))
            return 1 + options.datatype.size(value, registry)
        return 1
    }

    override fun serialize(value: Any?, buffer: ByteBuffer, registry: IRegistry) {
        buffer.put(if (!Objects.isNull(value)) 1.toByte() else 0.toByte())
        Optional.ofNullable(value).ifPresent { options.datatype.serialize(it, buffer, registry) }
    }

    override fun deserialize(buffer: ByteBuffer, registry: IRegistry): Any? {
        if (buffer.get() == 0.toByte())
            return null
        return options.datatype.deserialize(buffer, registry)
    }

    data class OptionTypeOptions(val datatype: ISerializable)

}

File: dynabuffers-python/dynabuffers/ast/datatype/OptionType.py
from dynabuffers.api.ISerializable import ISerializable


class OptionTypeOptions:
    def __init__(self, datatype: ISerializable):
        self.datatype = datatype


class OptionType(ISerializable):

    def __init__(self, options: OptionTypeOptions):
        self.options = options

    def size(self, value, registry):
        return 1 + (0 if value is None else self.options.datatype.size(value, registry))

    def serialize(self, value, buffer, registry):
        buffer.put(bytes([0 if value is None else 1]))
        if value is not None:
            self.options.datatype.serialize(value, buffer, registry)

    def deserialize(self, buffer, registry):
        if buffer.get() == 0:
            return None
        return self.options.datatype.deserialize(buffer, registry)
